include/pybind11/pybind11.h

```diff
@@ -917,7 +917,7 @@ class generic_type : public object {
     PYBIND11_OBJECT_DEFAULT(generic_type, object, PyType_Check)
 protected:
     void initialize(const type_record &rec) {
-        if (rec.scope && hasattr(rec.scope, rec.name))
+        if (rec.scope && hasattr(rec.scope, "__dict__") && rec.scope.attr("__dict__").contains(rec.name))
```

Comment:
Potentially, we could drop `hasattr(rec.scope, "__dict__")`, if we're happy throwing an exception when the user tries to use a scope that doesn't have a `__dict__` attribute?

Reply:
The idea of throwing an exception in this situation worries me, a lot.

I'm surprised to see the strong runtime performance concerns from both Yannick and Boris. Is it correct that this code only runs when a pybind11 module is imported? Is that so slow that people complain already? What's the incremental extra runtime overhead? My first guess is it's no measurable, given all the other things happening at import time (files are opened, dlls loaded), but maybe I'm overlooking something?

Reply:
> I'm surprised to see the strong runtime performance concerns from both Yannick and Boris.

It's not really a concern. It is only run once at import time, so it hardly matter.

> What's the incremental extra runtime overhead?

I wasn't measuring `import whatever`, but rather this specific patch in isolation.

```
>>> timeit.timeit(stmt='B().__dict__.get("X")', setup='class A:\n  def X(self):pass\nclass B(A):pass')
0.17064448199380422
>>> timeit.timeit(stmt='hasattr(B(), "X")', setup='class A:\n  def X(self):pass\nclass B(A):pass')
0.12798299700079951
```

Again, this is executed exactly once. It really shouldn't be a concern. I just wanted to satisfy my own curiosity.

Reply:
Not too worried about it, either! Just want to point it out.
The reasoning is/was, if `hasattr(rec.scope, "__dict__")` fails, most likely the later `scope.attr(name) = *this;` will also fail/throw an exception?

Reply:
I think I misunderstood this code initially (github is helpfully hiding the next line...). Would this be logically equivalent?
```
if (rec.scope && hasattr(rec.scope, rec.name) && hasattr(rec.scope, "__dict__") && rec.scope.attr("__dict__").contains(rec.name))
```
Now suspecting: the existing `hasattr` returns true, but we want to raise an error only if we find that the attribute is defined in `__dict__`?

Reply:
Hmmm, that's an interesting one! Yes, it might be better, this way.

You'll have figured out by now, but this is the failing condition, btw. This is when pybind11 throws an error.
I also figured out the reason it throws an error: this is the only place the created class/enum type gets stored. In the internals, only a (non-owning) `PyObject*` gets stored. If a pybind11 class would get overwritten and get garbage collected, `detail::internals`'s `type_map<type_info *> registered_types_cpp; // std::type_index -> pybind11's type information` map a C++ type to a Python type that doesn't exist anymore

Reply:
Thinking out loud (it's getting late on this side of the ocean, so I might be talking nonsense):

Would `if (rec.scope && hasattr(rec.scope, rec.name) && (!hasattr(rec.scope, "__dict__") || rec.scope.attr("__dict__").contains(rec.name)))` even be better?

I.e.:
- if we have no `scope`: weird, but OK, nevermind, we can't override anything
- if `hasattr` is false: fine, not problem, we can `setattr` the new class
- if `hasattr` is true, and we don't have a `__dict__`: aaaargh, something weird is going on in this type; and the `getattr` (and probably `setattr` as well) mechanism has been overwritten; so, custom case: go ahead, but you are probably specialized enough to find a way around it
- if `hasattr` is true, and we do have a `__dict__`: check if the dict contains it, if not, we probably found it in a parent class's scope

Reply:
(Writing that out: yeah, maybe the mix of boolean expressions could be rewritten more clearly, indeed)

Reply:
> we probably found it in a parent class's scope

The "probably" worries me a little.
This is beginning to look tricky.
Something totally crazy, with a vague hope that'll lead us somewhere good:

existing_obj = getattr(scope, name, None)
existing_refcount = getrefcount(existing_obj)
new_obj = setattr(scope, name, nested_obj)
if existing_obj is not None and getrefcount(existing_obj) < existing_refcount:
  raise

not good: makes assumptions about the refcounting system.
but the idea is to figure out if we actually squash an existing object (and create dangling internal pointers IIUC).

Reply:
> > we probably found it in a parent class's scope
> 
> The "probably" worries me a little.

Or somewhere else, but yes, it worries me as well.
Then again, for sane scope types, this is OK, I would say. pybind11 has a lot of assumptions where "By default and in sane behavior, this is OK; if you do crazy stuff, you're hopefully wise enough to know what dangers there are."

The refcounting seems dangerous, indeed. It should work, but it feels fishy. And PyPy might not find it as nice, either (since it's basically a CPython implementation detail).

Reply:
An other option is to make things overwriteable by keeping the classes somewhere else alive (hard to avoid leaks?) ? (Or, maybe even worse - because registered classes could "suddenly" become not convertible anymore - remove the `PyObject*` references from the internals when a `class_`-constructed Python type gets gc'd?)

Reply:
I totally agree with you. I was wondering about "pybind11 has a lot of assumptions", you just explained it, thanks! I agree with that, too. Your four-bullet logic makes sense AFAICT. If it's compatible with the general pybind11 philosophy, I think all we need is good hard tests (maybe complete already?) and good comments in the sources. Idea: move the logic into a separate function, so we can add comprehensive comments without cluttering the bigger context where the code lives right now?

Reply:
> An other option is to make things overwriteable by keeping the classes somewhere else alive (hard to avoid leaks?) ? (Or, maybe even worse - because registered classes could "suddenly" become not convertible anymore - remove the `PyObject*` references from the internals when a `class_`-constructed Python type gets gc'd?)

(side remark: I submitted my previous comment before I saw this.)

I'd venture to a more sprawling change like this only after deciding for sure your current local solution is infeasible.

Reply:
OK, I'll see if I can clean this up!

> I totally agree with you. I was wondering about "pybind11 has a lot of assumptions", you just explained it, thanks! I agree with that, too.

One last argument against that/worry: before, we were too strict on not wanting to overwrite an attribute (we wouldn't allow it even if it should be possible; only if `__getaatr__` and `__setattry__` are completely out of sync, we did something that could result in a crash). Now, we'd be slightly too lenient: we do allow it in some weird case where there's a dict but `__setattr__` doesn't overwrite in the dict (which could  get some pybind11 type instance garbage collected and cause a segfault).

Reply:
> - if `hasattr` is true, and we don't have a `__dict__`: aaaargh, something weird is going on in this type; and the getattr (and probably setattr as well) mechanism has been overwritten;

This is the case that worries me. Would this PR break this case?

Reply:
Potentially, yes. Because the previous `py::class_` would be replaced, could (maybe) get garbage collected, and pybind11 would hold a bad `PyObject*` pointer to the old class.

Reply:
Not sure how to solve this, though. Checking `hasattr`/`getattr` is too wide (i.e., checks base classes), while `__dict__` is too strict (would miss the crazy case where `__getattr__`/`__setattr__`, or `__getattribute__`/`__setattribute__`, would be overwritten).

Reply:
Maybe we shouldn't fix it? If it's this hard to do it right, perhaps we shouldn't "fix" it and break things that currently work.

Reply:
Interesting suggestion, yes. I must say the current situation feels wrong, though, where it matters in which order you register a class first in the base class or first in a derived class.

And the above shouldn't break too much. there's a thing where, if `getattr` and `setattr` are overwritten, things might already be broken anyway. In any case, the `scope` should be a class or a module, no?

Reply:
> In any case, the `scope` should be a class or a module, no?

Do you want the good news or the bad news? The bad news?

```
>>> def f():print('f')
...
>>> def g():print('g')
...
>>> f.g = g
>>> del g
>>> f.g()
g
>>> f()
f
```

The good news is that using functions as scopes isn't allowed with pybind. I tried. I really really tried to make pybind accept a function object as a scope.

Reply:
I'm indeed not sure if this is good or bad news :-/